### PR TITLE
WIP: Only ask the terminal for bg if unset

### PR DESCRIPTION
**Pull request by blueyed**
Ref: https://github.com/neovim/neovim/issues/11393

### CODE REVIEW

**Comment by justinmk** [comment]
This doesn't really fix https://github.com/neovim/neovim/issues/11393 , does it? And this makes the bg color detection almost useless since many colorschemes mindlessly set 'bg', IIRC.

**Comment by blueyed** [comment]
> This doesn't really fix #11393 , does it?

Yes, it does not fix it, but only is related.

> And this makes the bg color detection almost useless since many colorschemes mindlessly set 'bg', IIRC.

Good point.  OTOH `set bg` would still override the detected one then, wouldn't it?

This was just meant as an experiment to see how it goes.  The problematic part is doing this after the config is loaded, and there might be a better place/way to do it.

**Comment by justinmk** [comment]
Since this makes the logic/workflow more complicated I would say let's skip it.

**Comment by blueyed** [comment]
Yeah, it needs to be improved/fixed in general, but I think the approach is good: when `bg` is set we can skip this altogether (providing a workaround / optimization for when it is not needed in the first place).
What would be a good place to do it though?  Where is the initial configuration parsed/handled?

**Comment by justinmk** [comment]
> Where is the initial configuration parsed/handled?

Initial configuration of what?
 

> The problematic part is doing this after the config is loaded, and there might be a better place/way to do it.

 `VimEnter` could be used: https://github.com/neovim/neovim/blob/df90d25573ec94e1cd784cc8f7199f5be33477a1/src/nvim/tui/input.c#L425

**Comment by blueyed** [comment]
> Initial configuration of what?

I've meant startup scripts (https://github.com/neovim/neovim/blob/df90d25573ec94e1cd784cc8f7199f5be33477a1/src/nvim/main.c#L373-L374)

> `VimEnter` could be used:

Yeah, but I guess it is too late then likely to ask for the default bg.

Pushed some fixup, would have to look closer at existing failures etc.

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -1230,6 +1228,17 @@ static void show_termcap_event(void **argv)
   verbose_stop();  // flush now
 }
 
+/// Ask the terminal to send us the background color.
+static void ask_for_terminal_bg(UI *ui)
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: Let's use consistent names where possible.

```suggestion
static void request_terminal_bg(UI *ui)
```

**Comment by justinmk** [comment]
LGTM after name nit. I never cared much about this feature, I'd much rather just always set `bg=dark` :P

File: src/nvim/tui/tui.c [hunk 2]
```diff
@@ -1230,6 +1228,17 @@ static void show_termcap_event(void **argv)
...
+{
+  if (!option_was_set("bg") && !tui_is_stopped(ui)) {
```
Comment (erw7) [inline src/nvim/tui/tui.c:1234]: Please do not call the function that accesses the global variables of the main thread from the TUI thread.

https://github.com/neovim/neovim/blob/111d04fde64f8be8409cd3b1b96886aee2ac8ff5/src/nvim/tui/tui.c#L425

Accessing the global variables of the main thread from the TUI thread after this `CONTINUE(bridge);` may result in a data race.

**Comment by erw7** [comment]
I don't think this change will work because `ask_for_terminal_bg()` isn't always called after the main thread's `source_startup_scripts()`, etc. Perhaps the only appropriate place to call `ask_for_terminal_bg()` is before the next line.

https://github.com/neovim/neovim/blob/111d04fde64f8be8409cd3b1b96886aee2ac8ff5/src/nvim/tui/tui.c#L1068

**Pull request closed without merging**